### PR TITLE
feat: OpenAI-Compatible models, completions, chat/completions

### PR DESCRIPTION
**Pull request by bbrowning**
# What does this PR do?

This stubs in some OpenAI server-side compatibility with three new endpoints:

/v1/openai/v1/models
/v1/openai/v1/completions
/v1/openai/v1/chat/completions

This gives common inference apps using OpenAI clients the ability to talk to Llama Stack using an endpoint like
http://localhost:8321/v1/openai/v1 .

The two "v1" instances in there isn't awesome, but the thinking is that Llama Stack's API is v1 and then our OpenAI compatibility layer is compatible with OpenAI V1. And, some OpenAI clients implicitly assume the URL ends with "v1", so this gives maximum compatibility.

The openai models endpoint is implemented in the routing layer, and just returns all the models Llama Stack knows about.

The following providers should be working with the new OpenAI completions and chat/completions API:
* remote::anthropic (untested)
* remote::cerebras-openai-compat (untested)
* remote::fireworks (tested)
* remote::fireworks-openai-compat (untested)
* remote::gemini (untested)
* remote::groq-openai-compat (untested)
* remote::nvidia (tested)
* remote::ollama (tested)
* remote::openai (untested)
* remote::passthrough (untested)
* remote::sambanova-openai-compat (untested)
* remote::together (tested)
* remote::together-openai-compat (untested)
* remote::vllm (tested)

The goal to support this for every inference provider - proxying directly to the provider's OpenAI endpoint for OpenAI-compatible providers. For providers that don't have an OpenAI-compatible API, we'll add a mixin to translate incoming OpenAI requests to Llama Stack inference requests and translate the Llama Stack inference responses to OpenAI responses.

This is related to #1817 but is a bit larger in scope than just chat completions, as I have real use-cases that need the older completions API as well.

### vLLM

```
VLLM_URL="http://localhost:8000/v1" INFERENCE_MODEL="meta-llama/Llama-3.2-3B-Instruct" llama stack build --template remote-vllm --image-type venv --run

LLAMA_STACK_CONFIG=http://localhost:8321 INFERENCE_MODEL="meta-llama/Llama-3.2-3B-Instruct" python -m pytest -v tests/integration/inference/test_openai_completion.py --text-model "meta-llama/Llama-3.2-3B-Instruct"
```

### ollama
```
INFERENCE_MODEL="llama3.2:3b-instruct-q8_0" llama stack build --template ollama --image-type venv --run

LLAMA_STACK_CONFIG=http://localhost:8321 INFERENCE_MODEL="llama3.2:3b-instruct-q8_0" python -m pytest -v tests/integration/inference/test_openai_completion.py --text-model "llama3.2:3b-instruct-q8_0"
```

## Documentation

Run a Llama Stack distribution that uses one of the providers mentioned in the list above. Then, use your favorite OpenAI client to send completion or chat completion requests with the base_url set to http://localhost:8321/v1/openai/v1 . Replace "localhost:8321" with the host and port of your Llama Stack server, if different.

### CODE REVIEW

**Comment by bbrowning** [comment]
This is just a draft that I'll keep pushing on a bit to see how hard it would be to get OpenAI compatible completion and chat completion endpoints across all existing providers. Most of the work will come in the translation layer for the providers that don't use OpenAI clients internally.

This also paves the way to implementing OpenAI Responses API compatibility, but starting with Models, Completions, and Chat Completions first for the sake of maximum compatibility with existing clients in the wild.

**Review by terrytangyuan — COMMENTED** [review]
Thanks for the head start! Keep me posted on how this goes

File: llama_stack/apis/inference/inference.py [hunk 1]
```diff
@@ -564,3 +775,101 @@ async def embeddings(
         :returns: An array of embeddings, one for each content. Each embedding is a list of floats. The dimensionality of the embedding is model-specific; you can check model metadata using /models/{model_id}
         """
         ...
+
+    @webmethod(route="/openai/v1/completions", method="POST")
+    async def openai_completion(
+        self,
+        model: str,
+        prompt: str,
```
Comment (RobGeada) [inline llama_stack/apis/inference/inference.py:783]: The `prompt` field of the `/completions` endpoint  is a `string` or `array`, where the array is ["encoded as a string, array of strings, array of tokens, or array of token arrays."](https://platform.openai.com/docs/api-reference/completions/create#completions-create-prompt)- with this implementation you can't yet use arrays in the completions request.

Comment (bbrowning) [reply llama_stack/apis/inference/inference.py]: Good catch! I already had some in-progress changes that fix this as I worked on supporting the OpenAI server endpoints for non-OpenAI providers, so pushing some of that work a bit earlier than planned to go ahead and add the prompt as a string or array bits.

**Comment by bbrowning** [comment]
I'll start on some basic tests here, but note that I can't really write the typical inference integration tests until there's a release of the llama stack client python library with these API changes, since our integration tests use that to hit the inference endpoint.

**Comment by bbrowning** [comment]
Nevermind my comment about not being able to write integration tests yet. It dawned on me that I can of course just use an OpenAI client to test these new APIs, so I've started adding those.

**Comment by saichandrapandraju** [comment]
Is it possible to support [extra_body](https://github.com/openai/openai-python/blob/main/src/openai/resources/completions.py#L619) so that we can pass parameters that are not part of OpenAI API but others like vLLM specifically. 

eg., vLLM has a sampling parameter called [prompt_logprobs](https://docs.vllm.ai/en/latest/api/inference_params.html#:~:text=in%20the%20response.-,prompt_logprobs,-%E2%80%93%20Number%20of%20log) to return logprobs for prompt tokens ('[logprobs](https://docs.vllm.ai/en/latest/api/inference_params.html#:~:text=can%20be%20generated-,logprobs,-%E2%80%93%20Number%20of%20log)' parameter returns only for generated tokens). OpenAI has no 'prompt_logprobs' parameter, so I'm using `extra_body={'prompt_logprobs': 1}` to send this to vLLM server. 

This would require changes to response class to support prompt_logprobs. vLLM returns these prompt logprobs with 'prompt_logprobs' key in the choices list. Below is a sample response for the input prompt - `Dave lives in Palm Coast, FL and is a lawyer. His personal interests include` and vLLM sampling params `{
    "temperature":0,
    "prompt_logprobs": 1
}`

```
{'id': 'cmpl-5b44f8444e20400f837777b6b5c6f132',
 'object': 'text_completion',
 'created': 1743818276,
 'model': 'opt',
 'choices': [{'index': 0,
   'text': ' the law, politics, and business. He is a member of the Palm Coast',
   'logprobs': None,
   'finish_reason': 'length',
   'stop_reason': None,
   'prompt_logprobs': [None,
    {'33857': {'logprob': -13.873104095458984,
      'rank': 6057,
      'decoded_token': 'Dave'},
     '100': {'logprob': -1.4258383512496948, 'rank': 1, 'decoded_token': 'I'}},
    {'1074': {'logprob': -10.792854309082031,
      'rank': 2880,
      'decoded_token': ' lives'},
     '219': {'logprob': -2.0438311100006104, 'rank': 1, 'decoded_token': 'y'}},
    {'11': {'logprob': -0.4757663607597351,
      'rank': 1,
      'decoded_token': ' in'}},
    {'7929': {'logprob': -7.719001770019531,
      'rank': 295,
      'decoded_token': ' Palm'},
     '5': {'logprob': -2.083259344100952, 'rank': 1, 'decoded_token': ' the'}},
    {'2565': {'logprob': -4.589813232421875,
      'rank': 4,
      'decoded_token': ' Coast'},
     '2467': {'logprob': -0.6835634708404541,
      'rank': 1,
      'decoded_token': ' Beach'}},
    {'6': {'logprob': -0.5932450890541077, 'rank': 1, 'decoded_token': ','}},
    {'8854': {'logprob': -1.4891626834869385,
      'rank': 1,
      'decoded_token': ' FL'}},
    {'8': {'logprob': -1.737820029258728, 'rank': 3, 'decoded_token': ' and'},
     '4': {'logprob': -0.886257529258728, 'rank': 1, 'decoded_token': '.'}},
    {'16': {'logprob': -1.8477180004119873,
      'rank': 1,
      'decoded_token': ' is'}},
    {'10': {'logprob': -1.1920616626739502, 'rank': 1, 'decoded_token': ' a'}},
    {'2470': {'logprob': -6.246567726135254,
      'rank': 79,
      'decoded_token': ' lawyer'},
     '3562': {'logprob': -2.969223737716675,
      'rank': 1,
      'decoded_token': ' retired'}},
    {'4': {'logprob': -1.4258630275726318, 'rank': 1, 'decoded_token': '.'}},
    {'832': {'logprob': -3.3757128715515137,
      'rank': 4,
      'decoded_token': ' His'},
     '91': {'logprob': -0.8600877523422241,
      'rank': 1,
      'decoded_token': ' He'}},
    {'1081': {'logprob': -5.170414924621582,
      'rank': 32,
      'decoded_token': ' personal'},
     '173': {'logprob': -2.475102424621582,
      'rank': 1,
      'decoded_token': ' work'}},
    {'3168': {'logprob': -3.6241774559020996,
      'rank': 6,
      'decoded_token': ' interests'},
     '301': {'logprob': -2.0382399559020996,
      'rank': 1,
      'decoded_token': ' life'}},
    {'680': {'logprob': -0.6599060893058777,
      'rank': 1,
      'decoded_token': ' include'}}]}],
 'usage': {'prompt_tokens': 17,
  'total_tokens': 33,
  'completion_tokens': 16,
  'prompt_tokens_details': None}}
```

**Comment by bbrowning** [comment]
@saichandrapandraju I've wired in initial support for extra_body params for `prompt_logprobs` and `guided_choice` (because I needed this one for my own use-case) into this, along with a basic integration test for each that passes for me in vllm.

Just an implementation note, on the server-side these are top-level parameters I have to add to the `openai_completion` method because those `extra_body` parameters on the client-side come in as top-level body parameters when the actual request hits the server. That's not relevant to clients, but for implementing any additional `extra_body` parameters for all the backend providers we just have to hoist those up to optional parameters at the top level of the Inference API's `openai_completion` method.

File: llama_stack/providers/remote/inference/vllm/vllm.py [hunk 2]
```diff
@@ -440,8 +440,17 @@ async def openai_completion(
         temperature: Optional[float] = None,
         top_p: Optional[float] = None,
         user: Optional[str] = None,
+        guided_choice: Optional[List[str]] = None,
+        prompt_logprobs: Optional[int] = None,
     ) -> OpenAICompletion:
         model_obj = await self._get_model(model)
+
+        extra_body: Dict[str, Any] = {}
+        if prompt_logprobs:
```
Comment (saichandrapandraju) [inline llama_stack/providers/remote/inference/vllm/vllm.py:449]: For `prompt_logprobs`, 0 (zero) is a valid input. Can we do 'not None' and '>=0' check please?

Comment (bbrowning) [reply llama_stack/providers/remote/inference/vllm/vllm.py]: Yes, of course. I pushed that change and expanded the existing test I added for prompt_logprobs to cover testing a value of 0 as well. Thanks!

**Comment by ashwinb** [comment]
This is fantastic. Let me / @ehhuang know how we can help here. I will take a look at this in detail tonight or tomorrow.

**Comment by bbrowning** [comment]
I updated the PR description with the list of providers that may be working, marking which ones I've tested (remote::vllm and remote::ollama) and which ones I haven't. I also updated the commands I use to run the currently basic test suite against those providers, so that others can see how they'd run the tests against other providers.

Initial conversion of non-stream completion API calls to Llama Stack infererence calls is done as a fallback option for providers that don't natively support OpenAI APIs. I'll add in the same non-streaming fallback logic for chat completions soon. Feedback on whether we want to do this translation and extend it to also support streaming or whether we'd just prefer to raise an error stating that OpenAI APIs are unsupported for specific providers would be useful. For now, I'm erroring on the side of trying to make things Just Work for all inference providers, but that may not be trivial.

Also, I haven't yet had a chance to wire up all the providers that do support native OpenAI APIs, and will work on that. This includes databricks, nvidia, runpod, and sambanova specifically. Those are already using OpenAI clients so it should be quite trivial to wire them up just like I did for vllm, ollama, and together.ai.

File: llama_stack/apis/inference/inference.py [hunk 3]
```diff
@@ -564,3 +775,105 @@ async def embeddings(
...
+        parallel_tool_calls: Optional[bool] = None,
+        presence_penalty: Optional[float] = None,
+        response_format: Optional[Dict[str, str]] = None,
+        seed: Optional[int] = None,
+        stop: Optional[Union[str, List[str]]] = None,
+        stream: Optional[bool] = None,
+        stream_options: Optional[Dict[str, Any]] = None,
+        temperature: Optional[float] = None,
+        tool_choice: Optional[Union[str, Dict[str, Any]]] = None,
+        tools: Optional[List[Dict[str, Any]]] = None,
```
Comment (ehhuang) [inline llama_stack/apis/inference/inference.py:848]: should we have define a tool type for this?

File: tests/integration/inference/test_openai_completion.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (ehhuang) [inline tests/integration/inference/test_openai_completion.py:1]: I recently added a test suite to test OpenAI compat endpoints: https://github.com/meta-llama/llama-stack/blob/main/tests/verifications/README.md
We should move new test from here over there to consolidate if possible.

File: llama_stack/apis/inference/inference.py [hunk 5]
```diff
@@ -564,3 +775,105 @@ async def embeddings(
...
+        top_logprobs: Optional[int] = None,
+        top_p: Optional[float] = None,
+        user: Optional[str] = None,
+    ) -> OpenAIChatCompletion:
```
Comment (ehhuang) [inline llama_stack/apis/inference/inference.py:852]: Is this correctly typed for streaming?

File: llama_stack/apis/inference/inference.py [hunk 6]
```diff
@@ -564,3 +775,105 @@ async def embeddings(
         :returns: An array of embeddings, one for each content. Each embedding is a list of floats. The dimensionality of the embedding is model-specific; you can check model metadata using /models/{model_id}
         """
         ...
+
+    @webmethod(route="/openai/v1/completions", method="POST")
+    async def openai_completion(
```
Comment (ehhuang) [inline llama_stack/apis/inference/inference.py:780]: I wonder if we should have this under apis/openai/ so that OpenAI related things are in one place.

**Review by ehhuang — COMMENTED** [review]
Great start!

File: llama_stack/apis/inference/inference.py [hunk 7]
```diff
@@ -442,6 +442,217 @@ class EmbeddingsResponse(BaseModel):
...
+    """
+
+    finish_reason: str
+    text: str
+    index: int
+    logprobs: Optional[OpenAIChoiceLogprobs] = None
+
+
+@json_schema_type
+class OpenAICompletion(BaseModel):
```
Comment (ehhuang) [inline llama_stack/apis/inference/inference.py:639]: Should we just import from openai.types.chat as we did in openai_compat.py?

Comment (bbrowning) [reply llama_stack/apis/inference/inference.py:639]: I actually started with that. However, the API codegen wasn't able to successfully run with those types. I don't recall the exact errors now, but I can try an example out in a bit just to document what the actual issue was there. A secondary concern would be whether we want direct control over the public-facing API of Llama Stack or whether we want to let new versions of the OpenAI python client impact our API surface.

File: llama_stack/apis/inference/inference.py (hunk 6 above)
Comment (bbrowning) [reply llama_stack/apis/inference/inference.py:780]: That's reasonable, and I went back-and-forth a bit here myself. I put the OpenAI models API endpoint under our models.py file and the OpenAI inference endpoints under our inference.py file simply because they mapped nicely to existing constructs.  But, I don't have a strong preference there.

File: llama_stack/apis/inference/inference.py (hunk 5 above)
Comment (bbrowning) [reply llama_stack/apis/inference/inference.py:852]: No, it's not. The type doesn't cover the streaming case at all, so even though streaming works in practice with the API as-is when used from OpenAI clients the typing for streaming isn't handled yet.

File: llama_stack/apis/inference/inference.py (hunk 7 above)
Comment (bbrowning) [reply llama_stack/apis/inference/inference.py:639]: Here's an example of the kinds of errors the API spec codegen throws when using any of the OpenAI python client's types in our API:

```
Traceback (most recent call last):
  File "/Users/bbrowning/.pyenv/versions/3.10.16/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/bbrowning/.pyenv/versions/3.10.16/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/generate.py", line 91, in <module>
    fire.Fire(main)
  File "/Users/bbrowning/.cache/uv/archive-v0/fsDRrIpMBoxSdg6tsSQLY/lib/python3.10/site-packages/fire/core.py", line 135, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/Users/bbrowning/.cache/uv/archive-v0/fsDRrIpMBoxSdg6tsSQLY/lib/python3.10/site-packages/fire/core.py", line 468, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/Users/bbrowning/.cache/uv/archive-v0/fsDRrIpMBoxSdg6tsSQLY/lib/python3.10/site-packages/fire/core.py", line 684, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/generate.py", line 55, in main
    spec = Specification(
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/utility.py", line 29, in __init__
    self.document = generator.generate()
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 781, in generate
    operation = self._build_operation(op)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 691, in _build_operation
    responses = response_builder.build_response(response_options)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 374, in build_response
    responses[status_code] = self._build_response(
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 393, in _build_response
    content=self.content_builder.build_content(response_type, examples),
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 216, in build_content
    return {media_type: self.build_media_type(item_type, examples)}
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 221, in build_media_type
    schema = self.schema_builder.classdef_to_ref(item_type)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 135, in classdef_to_ref
    type_schema = self.classdef_to_schema(typ)
  File "/Volumes/SourceCode/llama-stack/docs/openapi_generator/pyopenapi/generator.py", line 116, in classdef_to_schema
    type_schema, type_definitions = self.schema_generator.classdef_to_schema(typ)
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 612, in classdef_to_schema
    types_defined[sub_name] = self._type_to_schema_with_lookup(sub_type)
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 569, in _type_to_schema_with_lookup
    type_schema = self.type_to_schema(data_type, force_expand=True)
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 321, in type_to_schema
    return self._type_to_schema(data_type, force_expand, json_schema_extra) | common_info
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 518, in _type_to_schema
    property_def = self.type_to_schema(property_type, json_schema_extra=json_schema_extra)
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 321, in type_to_schema
    return self._type_to_schema(data_type, force_expand, json_schema_extra) | common_info
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/schema.py", line 495, in _type_to_schema
    for property_name, property_type in get_class_properties(typ):
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/inspection.py", line 571, in get_class_properties
    resolved_hints = get_resolved_hints(typ)
  File "/Volumes/SourceCode/llama-stack/llama_stack/strong_typing/inspection.py", line 557, in get_resolved_hints
    return typing.get_type_hints(typ, include_extras=True)
  File "/Users/bbrowning/.pyenv/versions/3.10.16/lib/python3.10/typing.py", line 1833, in get_type_hints
    value = _eval_type(value, base_globals, base_locals)
  File "/Users/bbrowning/.pyenv/versions/3.10.16/lib/python3.10/typing.py", line 327, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/Users/bbrowning/.pyenv/versions/3.10.16/lib/python3.10/typing.py", line 694, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'ClassVar' is not defined
```

It's probably solvable, but something about how the OpenAI types use `ClassVar` isn't liked by the strong_typing code in Llama Stack.

**Comment by ehhuang** [comment]
FYI I'm building on top of this to enable OAI compat for meta-reference inference

**Comment by ashwinb** [comment]
Chatted with Eric and Raghu. We feel comfortable merging this in, and iterating on it in origin/main both to indicate to folks that this is imminent and also avoid unnecessary pain for @bbrowning. Others can also potentially jump in. 

Unless someone objects in the next hour or so, I will do so. We will try to get a release going later (or over the weekend) potentially as well.

**Review by ashwinb — APPROVED** [review]

**Comment by bbrowning** [comment]
Awesome, thanks for the quick turnaround! Now we should be able to get a lot more real-world testing of this to work out the edge cases.
